<a href="https://colab.research.google.com/github/KILjungjoon/Language-Statistics/blob/main/Breusch_Godfrey_Test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

https://www.statology.org/breusch-godfrey-test-python/

H0 (null hypothesis): There is no autocorrelation at any order less than or equal to p.

HA (alternative hypothesis): There exists autocorrelation at some order less than or equal to p.

# Step 1: Create the Data

In [1]:
import pandas as pd 

#create dataset
df = pd.DataFrame({'x1': [3, 4, 4, 5, 8, 9, 11, 13, 14, 16, 17, 20],
                   'x2': [7, 7, 8, 8, 12, 4, 5, 15, 9, 17, 19, 19],
                    'y': [24, 25, 25, 27, 29, 31, 34, 34, 39, 30, 40, 49]})

#view first five rows of dataset
df.head()

x1  x2   y
0   3   7  24
1   4   7  25
2   4   8  25
3   5   8  27
4   8  12  29

# Step 2: Fit a Regression Model

In [2]:
import statsmodels.api as sm

#define response variable
y = df['y']

#define predictor variables
x = df[['x1', 'x2']]

#add constant to predictor variables
x = sm.add_constant(x)

#fit linear regression model
model = sm.OLS(y, x).fit()

/usr/local/lib/python3.8/dist-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


# Step 3: Perform the Breusch-Godfrey Test

In [3]:
import statsmodels.stats.diagnostic as dg

#perform Breusch-Godfrey test at order p = 3
print(dg.acorr_breusch_godfrey(model, nlags=3))

(8.703148274278197, 0.03350948731797309, 5.279672243902776, 0.04039805764950987)


* Test statistic X2 = 8.7031
* P-value = 0.0335
* Since this p-value is less than 0.05, we can reject the null hypothesis and conclude that autocorrelation exists among the residuals at some order less than or equal to 3.